File: game-finder/src/main/java/gamefinder/ProfilesManagement.java
package gamefinder;
/**
 * 
 * @Author: Andrew Skevington-Olivera
 * @Date: 16-4-23
 */


import java.util.ArrayList;

import org.bson.Document;
 
import com.mongodb.MongoClient;
import com.mongodb.client.MongoCollection;

/**
 * Profiles[]
 * getProfiles()
 * checkLogin()
 * returnProfileUsernames()
 * signUp()
 * addProfiles()
 * removeProfile()
 * accessUser()
 * returnProfiles[] ()
 * 
 * 
 */

public class ProfilesManagement {

    private ArrayList<Profile> profiles = new ArrayList<>();
    private GetDbCollection mongoDB = new GetDbCollection();
    private MongoClient mongoClient = null;
    private MongoCollection<Document> profileCollection = mongoDB.returnCollection("ProfilesDB", "Profiles", this.mongoClient);;

    ProfilesManagement( MongoClient mongoClient ){

        this.mongoClient = mongoClient;
        //this.profileCollection = mongoDB.returnCollection("ProfilesDB", "Profiles", this.mongoClient);

        for(Document doc : profileCollection.find() ){
            String username = doc.getString("Username");
            String pw = doc.getString("Password");
            addProfile( new Profile(username, pw, this.mongoClient) );
        }


    }


    public String signUp(String displayName, String username, String pw){

        for(Document doc : profileCollection.find() ){
            if( username.compareTo( doc.getString("Username") ) == 0) {
                return "False";     //False means that the username is already taken
            }
        }

        Profile newProfile = new Profile(displayName, username, pw, this.mongoClient);
        addProfile( newProfile );

        return "True";  //True means that the new profile was created and there wasn't the same username.
        
    }

    public String login(String username, String pw){

        for(Document doc : profileCollection.find() ) {
            if( (username.compareTo( doc.getString("Username") ) == 0) && (pw.compareTo( doc.getString("Password") ) == 0) ) {
                return "True";  //The login information was correct
            }
        }

        return "False";     //The login information wasn't correct
    }

    void addProfile(Profile profile){
        profiles.add(profile);
    }

    public Profile accessProfile(String username) { //Change the username to token for later

        for(Profile profile : this.profiles){
            if( username.compareTo( profile.getUsername() ) == 0){
                return profile;
            }
        }

        return null;
    }

    public ArrayList<Profile> returnProfiles(){
        return profiles;
    }


    
}

File: game-finder/src/backend/ProfilesManagement.ts
/**
 * The ProfilesManagement class handles the storage and management of an array of profiles. It primarily focuses on the backend functionalities
 * related to login, signup, and general profile management.
 * The ProfilesManagement class provides the following functions:
 * initialize(): Retrieves all the profiles stored in the database and initializes the Profile class for each profile,
 * allowing for efficient and organized information manipulation.
 * signIn(): Handles the creation of a new profile by checking if the username is available and storing the new profile in the database.
 * If the username is already taken, it returns a string indicating that the username is unavailable.
 * login(): Verifies the login credentials provided by a user and returns a boolean value indicating the success or failure of the login.
 * accessUser(): Given a username, it searches through the profile list for potential matches and returns the corresponding profile information.
 * returnProfileUsernames(): Returns an array containing all the usernames of the profiles stored in the ProfileList.
 * @Author Andrew SKevington-Olivera
 * 19-4-23
 */


import { MongoDB } from "./mongoDB";
import { Profile } from "./Profile";

export class ProfileManagement {

    public profileList = new Array();
    private db : MongoDB;

    constructor( db : MongoDB) {
        this.db = db;
        this.initialize();
    }

    /**
     * This creates a Profile Class for each of the different Profiles that're being stored in the Database.
     * It then adds it to the profileList array and makes sure all of the information is being transferred over from the
     * database into the Profile Object.
     */
    async initialize() {

        const collection = this.db.returnCollection("ProfilesDB", "Profiles");

        var usernameList = new Array();
        var pwList = new Array();

        await collection.find().forEach( function(myDoc: { Username: string; } ) { usernameList.push(myDoc.Username) }  );
        await collection.find().forEach( function(myDoc: { Password: string; } ) { pwList.push(myDoc.Password) } );

        for(var i = 0; i < usernameList.length; i++){
            let oldProfile = new Profile( usernameList[i], pwList[i], this.db);
            this.profileList.push( oldProfile );
        }
    }

    /**
     * This checks if the username that the user wants is taken or not, if the username isn't taken then a new Profile Obj
     * will be created and pushed to the array, and it'll be saved to the database.
     * @param displayName 
     * @param email 
     * @param username 
     * @param password 
     * @returns Returns true if username hasn't been chosen yet, false if username is taken.
     */
    async signIn(displayName : string, email : string, username : string, password : string) {

        const collection = this.db.returnCollection("ProfilesDB", "Profiles");
        const doc = await collection.findOne( {Username : username} );

        if(doc == null) {    //Username hasn't been chosen yet for the website
            let newProfile = new Profile(displayName, email, username, password, this.db);
            this.profileList.push( newProfile )
            console.log("A new profile should be in the process of being created");
            return true;
        }
        
        console.log( "The username was already being taken" );
        return false;   //The username is already taken
    }

    /**
     * Checks if the login information is correct and returns a bool to say true or false for the login credentials given.
     * @param username 
     * @param password 
     * @returns true is information is correct & false if it's wrong.
     */
    async login( username : string, password : string){

        const collection = this.db.returnCollection("ProfilesDB", "Profiles");
        const doc = await collection.findOne( {Username : username} );
        if(doc != null) {
            if( password === doc.Password) {
                return true;    //Information was correct.
            }
            else {
                return false;   //Password wasn't correct. User should try to enter their login information again.
            }
        }
        else{
            return false;   //Username wasn't in the database. User should try to enter their login information again.
        }
    }

    /**
     * Looks through all of the profiles in profileList and sees if any of the usernames matche up.
     * @param username
     * @returns Profile Obj with the same username.
     */
    public async accessUser(username : string) {

        for(var i = 0; i < this.profileList.length; i++) {
            if(username == this.profileList[i].returnUsername() ) {
                let copyProfile = this.profileList[i];  //Can probably just return this.profileList[i] later
                return copyProfile;
            }
        }
    }


    /**
     * This goes through each of the profiles in the profileList array and appends their username to an array
     * to return.
     * @returns usernames of all profiles in profileList array.
     */
    public returnProfileUsernames() {

        let usernameList = new Array();

        for(var i = 0; i < this.profileList.length; i++) {
            usernameList.push( this.profileList[i].username );
        }   

        return usernameList;
    }


}